Set RouteMTU for generic veth

sugangli opened the pull request:
Please ensure your pull request adheres to the following guidelines:

- [ ] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [ ] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [ ] Provide a title or release-note blurb suitable for the release notes.
- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [x] Thanks for contributing!

<!-- Description of change -->
The PR inherits from https://github.com/cilium/cilium/pull/21994.  Once `enable-route-mtu` is true, route MTU will be applied to every route inside the pod. Also resolved some minor conflict in the second commit.

aojea commented on plugins/cilium-cni/types/types.go:
```diff
@@ -22,16 +22,21 @@ import (
+	ChainingMode   string                 `json:"chaining-mode"`
+}
+
+type ChainingMode struct {
```
this type is not needed, is it?

sugangli replied:
Deleted.

aojea commented on plugins/cilium-cni/utils/utils.go:
```diff
@@ -0,0 +1,20 @@
+package utils
```
I think you need the header

```
// SPDX-License-Identifier: Apache-2.0
// Copyright Authors of Cilium
```

sugangli replied:
Done.

aojea commented on plugins/cilium-cni/chaining/generic-veth/generic-veth.go:
```diff
@@ -102,10 +110,33 @@ func (f *GenericVethChainer) Add(ctx context.Context, pluginCtx chainingapi.Plug
+		}
+
+		if pluginCtx.NetConf.EnableRouteMTU {
+			routes, err = netlink.RouteList(nil, netlink.FAMILY_V4)
```
only ipv4?

sugangli replied:
Added the support for Ipv6 as well.

tklauser commented on plugins/cilium-cni/chaining/generic-veth/generic-veth.go:
```diff
@@ -49,11 +51,17 @@ func (f *GenericVethChainer) Add(ctx context.Context, pluginCtx chainingapi.Plug
 		netNs                                               ns.NetNS
 	)
 
+	conf, err := ciliumcniutils.GetConfigFromCiliumAgent(cli)
```
Instead of factoring this out into `ciliumcniutils` and calling the Cilium agent OpenAPI call a second time to retrieve the agent config here, how about storing the result of the initial call in `cmdAdd` as `*models.DaemonConfigurationStatus` in `PluginContext` and then use it here? The agent config is not expected to change in between the calls.

sugangli replied:
Make sense.  I have updated accordingly.

sugangli committed:
Set RouteMTU for generic veth

This change take the tunnel overhead into account. Once
`enable-route-mtu` is true, route MTU will be applied to every route
inside the pod. If cilium-agent is not aware of any tunneling, then
route MTU will be equal to the link MTU.

Co-authored-by: Fei Su <sofat1989@126.com>
Signed-off-by: Sugang Li <sugangli@google.com>

aojea commented on plugins/cilium-cni/chaining/generic-veth/generic-veth.go:
```diff
@@ -49,11 +49,16 @@ func (f *GenericVethChainer) Add(ctx context.Context, pluginCtx chainingapi.Plug
 		netNs                                               ns.NetNS
 	)
 
+	if err != nil {
```
this does not make sense here, does it?

aojea commented on plugins/cilium-cni/main.go:
```diff
@@ -397,6 +397,8 @@ func cmdAdd(args *skel.CmdArgs) (err error) {
 	}
 
+	conf, err := getConfigFromCiliumAgent(c)
+
```
you moved the line bot not the err check

sugangli replied:
Ah... Updated.

aojea commented on plugins/cilium-cni/main.go:
```diff
@@ -440,7 +443,6 @@ func cmdAdd(args *skel.CmdArgs) (err error) {
 			args.IfName, args.Netns, err)
 	}
 
-	conf, err := getConfigFromCiliumAgent(c)
```
https://github.com/cilium/cilium/pull/26495/files#r1245744566

squeed reviewed (approved):
This looks good.

Please consider a follow-up PR that adds this value to the `cni` section in the Helm chart. This will then need to be plumbed through the entire configuration system, all the way down to the CNI file.
